typescript/connections.md

```diff
@@ -0,0 +1,151 @@
+import { Connection, WorkflowClient } from '@temporalio/client';
+
+const connection = await Connection.create({ address });
+const client = new WorkflowClient(connection.service, { namespace });
+```
+
+#### Pros
+
+- Explicit connection reuse
+
+#### Cons
+
+- Verbose
+- Confusing when to use each connection type
+
+### Implicit connection reuse with similar options
+
+```ts
+// worker.ts
+import { Worker } from '@temporalio/worker';
+
+// worker1 and worker2 share the same underlying connection based on given options
+const worker1 = await Worker.create({ connectionOptions: { address }, namespace, taskQueue: 'tq1' });
+const worker2 = await Worker.create({ connectionOptions: { address }, namespace, taskQueue: 'tq2' });
+
+// starter.ts
+import { WorkflowClient } from '@temporalio/client';
+
+const client = await WorkflowClient.create({ connectionOptions: { address }, namespace });
+```
+
+#### Pros
+
+- Short
+- Easy path from sample with local server to real world use case
+
+#### Cons
+
+- Implicit connection sharing is non-obvious
+- No way to opt out of connection sharing (can be addressed if needed)
+
+### Mix of options and explicit clients (recommended)
+
+Support a variant of both approaches listed above with no implicit connection reuse
+
+```ts
+// worker.ts
+import { NativeConnection, Worker } from '@temporalio/worker';
+
+// worker1 and worker2 do not share the same connection
+const worker1 = await Worker.create({ connectionOptions: { address }, namespace, taskQueue: 'tq1' });
+const worker2 = await Worker.create({ connectionOptions: { address }, namespace, taskQueue: 'tq2' });
+
+// worker3 and worker4 share the same connection
+const connection = await NativeConnection.create({ address });
+const worker3 = await Worker.create({ connection, namespace, taskQueue });
+const worker4 = await Worker.create({ connection, namespace, taskQueue });
+
+// starter.ts
+import { Connection, WorkflowClient } from '@temporalio/client';
+
+const client = await WorkflowClient.create({ connectionOptions: { address }, namespace });
+
+const connection = await Connection.create({ address });
+const client = new WorkflowClient({ connection, namespace });
+```
+
+### Python SDK approach
+
+```ts
+// worker.ts
+import { Worker, WorkflowClient } from '@temporalio/worker';
+// WorkflowClient is re-exported from the worker package
+
+// worker1 and worker2 share the same underlying connection based on given options
+const client = await WorkflowClient.create({ connectionOptions: { address }, namespace });
+const worker = await Worker.create({ client, namespace, taskQueue: 'tq1' });
+
+// starter.ts
+import { WorkflowClient } from '@temporalio/client';
+
+const client = await WorkflowClient.create({ connectionOptions: { address }, namespace });
+```
+
+#### Pros
+
+- One way to create clients
+- Explicit sharing model
+
+#### Cons
+
+- Connnection isn't really reused (not possible)
```

Comment:
Why is it not possible to reuse the connection? Is it because the client is pure TS gRPC and the worker is core? If we're not going to back the client by core, then that limitation exists for every client vs worker wrt connection reuse. 

But it doesn't mean connections can't be used _across workers_. You just store an internal, lazily created Core connection no the client when it's first needed. Then every worker created from that client can use the same connection.

The only "Con" I see with this approach if pure TS gRPC client remains is that for worker-only situations, you are creating/connecting via pure TS gRPC unnecessarily.

Reply:
If we default to using the Core client we could reuse connections, I'm debating whether we want to go that route.
The JS client has more features and flexibility so it'd be hard to give it up.

Reply:
Definitely don't need to give up the JS client. I just want to make sure that connection reuse across workers is clear.

Reply:
Yes, for sure possible across workers but not between a worker and a client.